### PR TITLE
Sdl wayland support

### PR DESCRIPTION
**Pull request by infapi00**
Two first commits are just utility/misc, but the other two get the sponza working on full wayland environments.

The easier way to test them on a environment with X, XWayland, etc is using the following envvar:

`export SDL_VIDEODRIVER=wayland`

I'm not fully happy with the last commit, so this Pull request would work as a discussion thread too.

### CODE REVIEW

File: framework/vkdf-init.cpp [hunk 1]
```diff
@@ -575,7 +489,7 @@ _init_swap_chain(VkdfContext *ctx)
    swap_chain_info.preTransform = present_transform;
    swap_chain_info.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
    swap_chain_info.imageArrayLayers = 1;
-   swap_chain_info.presentMode = present_mode;
+   swap_chain_info.presentMode = VK_PRESENT_MODE_FIFO_KHR;
```
Comment (itoral) [inline framework/vkdf-init.cpp]: NAK -- Mesa is not the only Vulkan implementation available even if it is the one we use :)

File: README [hunk 2]
```diff
@@ -91,4 +91,8 @@ generated shader files are re-generated again:
 $ make clean
 $ make -j
 
+Support on full wayland environments are only working using SDL.
```
Comment (itoral) [inline README]: I'd replace this with: "Wayland support is only available via SDL"

File: framework/vkdf-platform-sdl.cpp [hunk 3]
```diff
@@ -78,22 +78,41 @@ vkdf_platform_create_window(VkdfPlatform *platform,
...
+   switch(win_info.subsystem) {
+   case SDL_SYSWM_X11: {
+      VkXlibSurfaceCreateInfoKHR create_info;
+      create_info.sType = VK_STRUCTURE_TYPE_XLIB_SURFACE_CREATE_INFO_KHR;
+      create_info.pNext = NULL;
+      create_info.flags = 0;
+      create_info.dpy = win_info.info.x11.display;
+      create_info.window = win_info.info.x11.window;
+      res = vkCreateXlibSurfaceKHR(inst, &create_info, NULL, &platform->surface);
+
```
Comment (itoral) [inline framework/vkdf-platform-sdl.cpp]: Drop the empty line here.

File: framework/vkdf-platform-sdl.cpp [hunk 4]
```diff
@@ -78,22 +78,41 @@ vkdf_platform_create_window(VkdfPlatform *platform,
...
+   }
+   case SDL_SYSWM_WAYLAND: {
+      VkWaylandSurfaceCreateInfoKHR create_info;
+      create_info.sType = VK_STRUCTURE_TYPE_WAYLAND_SURFACE_CREATE_INFO_KHR;
+      create_info.pNext = NULL;
+      create_info.flags = 0;
+      create_info.display = win_info.info.wl.display;
+      create_info.surface = win_info.info.wl.surface;
+      res = vkCreateWaylandSurfaceKHR(inst, &create_info, NULL, &platform->surface);
+
```
Comment (itoral) [inline framework/vkdf-platform-sdl.cpp]: Same.

File: framework/vkdf-platform-sdl.cpp [hunk 5]
```diff
@@ -78,22 +78,41 @@ vkdf_platform_create_window(VkdfPlatform *platform,
...
+      break;
+   }
+   default:
+      vkdf_fatal("SDL win_info.subsystem not supported: %i\n", win_info.subsystem);
+   }
 
    if (res != VK_SUCCESS)
-      vkdf_fatal("Failed to create window surface");
+      vkdf_fatal("Failed to create window surface: %i", res);
```
Comment (itoral) [inline framework/vkdf-platform-sdl.cpp]: "Failed to create window surface. Error code: %i"

File: framework/vkdf-event-loop.cpp [hunk 6]
```diff
@@ -111,6 +113,20 @@ acquire_next_image(VkdfContext *ctx)
                                   NULL,
                                   &ctx->swap_chain_index);
 
+      if (sem_index != ctx->swap_chain_index) {
```
Comment (itoral) [inline framework/vkdf-event-loop.cpp:116]: The if body is not properly indented.

File: framework/vkdf-event-loop.cpp [hunk 7]
```diff
@@ -111,6 +113,20 @@ acquire_next_image(VkdfContext *ctx)
...
+	 // This happens when swap_chain_index goes back to zero
+	 // without going through the full swap chain
+	 assert(ctx->swap_chain_index == 0);
+
+	 // We swap the semaphore for the acquired image with the
+	 // temporary index to the one in the swap_chain_index
+	 // position
+
+	 VkSemaphore aux = ctx->acquired_sem[ctx->swap_chain_index];
```
Comment (itoral) [inline framework/vkdf-event-loop.cpp]: You can use the swap() C++ function for this.

**Comment by itoral** [comment]
I NAK'ed  the first patch, but the other 3 look good with the feedback I suggested addressed. About the last patch, it is not pretty, but since the kludge is contained within the one function it changes I think we can live with it as-is.

**Review by itoral — CHANGES_REQUESTED** [review]

**Comment by infapi00** [comment]
Applied all feedback.

File: framework/vkdf-platform-sdl.cpp (hunk 5 above)
Comment (itoral) [reply framework/vkdf-platform-sdl.cpp]: This was not fixed.

Comment (infapi00) [reply framework/vkdf-platform-sdl.cpp]: yes it was:

https://github.com/itoral/vkdf/pull/11/commits/929e4296b0934cae32b473503c8fe37b112c6a0f#diff-1462c1d4876365bf587fac0c7833ea6d8a32f5269a42cb20eaf96e7ec8ab6ddaR112

perhaps you were still looking at the old commits?

Comment (itoral) [reply framework/vkdf-platform-sdl.cpp]: Ah no, I was looking at a different line in the code, never mind.

**Pull request merged**